src/main/java/com/exadel/discountwebapp/userdiscount/qrcodegenerator/QRCodeService.java

```diff
@@ -14,26 +16,32 @@
 
     private final UserRepository userRepository;
     private final DiscountRepository discountRepository;
+    private final UserDiscountRepository userDiscountRepository;
 
     public QRCodeResponseVO qrCodeViewPage(Long userId, Long discountId) {
-        var user = userRepository.findById(userId)
-                .orElseThrow(() -> new EntityNotFoundException(User.class, "id", userId));
-        var discount = discountRepository.findById(discountId)
-                .orElseThrow(() -> new EntityNotFoundException(Discount.class, "id", discountId));
-
-        QRCodeResponseVO qrcodeResponse = new QRCodeResponseVO();
-        qrcodeResponse.setFirstName(user.getFirstName());
-        qrcodeResponse.setLastName(user.getLastName());
-        qrcodeResponse.setImageUrl(user.getImageUrl());
-
-        qrcodeResponse.setVendorTitle(discount.getVendor().getTitle());
-        qrcodeResponse.setVendorEmail(discount.getVendor().getEmail());
-
-        qrcodeResponse.setDiscountTitle(discount.getTitle());
-        qrcodeResponse.setPromocode(discount.getPromocode());
-        qrcodeResponse.setDiscountStartDate(discount.getStartDate());
-        qrcodeResponse.setDiscountExpirationDate(discount.getExpirationDate());
-
-        return qrcodeResponse;
+        UserDiscount.UserDiscountId userDiscountId = new UserDiscount.UserDiscountId(userId, discountId);
+        if (userDiscountRepository.existsById(userDiscountId)) {
+            var user = userRepository.findById(userId)
+                    .orElseThrow(() -> new EntityNotFoundException(User.class, "id", userId));
+            var discount = discountRepository.findById(discountId)
+                    .orElseThrow(() -> new EntityNotFoundException(Discount.class, "id", discountId));
+
+            QRCodeResponseVO qrcodeResponse = new QRCodeResponseVO();
+            qrcodeResponse.setFirstName(user.getFirstName());
+            qrcodeResponse.setLastName(user.getLastName());
+            qrcodeResponse.setImageUrl(user.getImageUrl());
+
+            qrcodeResponse.setVendorTitle(discount.getVendor().getTitle());
+            qrcodeResponse.setVendorEmail(discount.getVendor().getEmail());
+
+            qrcodeResponse.setDiscountTitle(discount.getTitle());
+            qrcodeResponse.setPromocode(discount.getPromocode());
+            qrcodeResponse.setDiscountStartDate(discount.getStartDate());
+            qrcodeResponse.setDiscountExpirationDate(discount.getExpirationDate());
+
+            return qrcodeResponse;
+        } else {
+            throw new EntityNotFoundException(UserDiscount.class, "id", userDiscountId);
+        }
```

Comment:
it is better to first check if userDiscount does not exist, and then throw exception

and then create response
in this way we will avoid "else" part

it is easier to read code in this way. 
firstly we validate data, and then process it

Reply:
changed